src/libxtp/segment.cc

```diff
@@ -0,0 +1,542 @@
+// be able to access it. Used for creating the ghost in PB corrected pairs.
+Segment::Segment(Segment *stencil)
+    : _id(stencil->getId()),
+      _name(stencil->getName() + "_ghost"),
+      _typ(stencil->getType()),
+      _top(NULL),
+      _mol(NULL),
+      _CoM(stencil->getPos()),
+      _has_e(false),
+      _has_h(false),
+      _has_s(false),
+      _has_t(false) {
+  _eMpoles.resize(5);
+
+  vector<Fragment *>::iterator fit;
+  for (fit = stencil->Fragments().begin(); fit < stencil->Fragments().end();
+       fit++) {
+
+    Fragment *newFrag = new Fragment(*fit);
+    this->AddFragment(newFrag);
+
+    vector<Atom *>::iterator ait;
+    for (ait = newFrag->Atoms().begin(); ait < newFrag->Atoms().end(); ait++) {
+      this->AddAtom(*ait);
+    }
+  }
+}
+
+Segment::~Segment() {
+
+  vector<Fragment *>::iterator fragit;
+  for (fragit = this->Fragments().begin(); fragit < this->Fragments().end();
+       fragit++) {
+    delete *fragit;
+  }
+  _fragments.clear();
+  _atoms.clear();
+
+  _eMpoles.clear();
+  _polarSites.clear();
+}
+
+void Segment::TranslateBy(const vec &shift) {
+
+  _CoM = _CoM + shift;
+
+  vector<Fragment *>::iterator fit;
+  for (fit = _fragments.begin(); fit < _fragments.end(); fit++) {
+
+    (*fit)->TranslateBy(shift);
+  }
+}
+
+void Segment::setHasState(bool yesno, int state) {
+
+  if (state == -1) {
+    _has_e = yesno;
+  } else if (state == +1) {
+    _has_h = yesno;
+  } else if (state == +2) {
+    _has_s = yesno;
+  } else if (state == +3) {
+    _has_t = yesno;
+  } else {
+    throw runtime_error(" ERROR CODE whe__00e11h__");
+  }
+}
+
+bool Segment::hasState(int state) {
+  bool result;
+  if (state == -1) {
+    result = _has_e;
+  } else if (state == +1) {
+    result = _has_h;
+  } else if (state == +2) {
+    result = _has_s;
+  } else if (state == +3) {
+    result = _has_t;
+  } else {
+    throw runtime_error(" ERROR CODE whe__00s11o__");
+  }
+  return result;
+}
+
+void Segment::setOcc(double occ, int e_h_s_t) {
+
+  if (e_h_s_t == -1) {
+    _occ_e = occ;
+  } else if (e_h_s_t == +1) {
+    _occ_h = occ;
+  } else if (e_h_s_t == +2) {
+    _occ_s = occ;
+  } else if (e_h_s_t == +3) {
+    _occ_t = occ;
+  } else {
+    throw runtime_error(" ERROR CODE whe__00s11o__");
+  }
+}
+
+double Segment::getOcc(int e_h_s_t) {
+  double result;
+  if (e_h_s_t == -1) {
+    result = _occ_e;
+  } else if (e_h_s_t == +1) {
+    result = _occ_h;
+  } else if (e_h_s_t == +2) {
+    result = _occ_s;
+  } else if (e_h_s_t == +3) {
+    result = _occ_t;
+  } else {
+    throw runtime_error(
+        " ERROR CODE whe__00s11o__");  // blabla what do I do here?
+  }
+  return result;
+}
+
+void Segment::setU_cC_nN(double dU, int state) {
+
+  if (state == -1) {
+    _U_cC_nN_e = dU;
+  } else if (state == +1) {
+    _U_cC_nN_h = dU;
+  } else {
+    throw runtime_error(" ERROR CODE whe__00u11a__");
+  }
+}
+
+void Segment::setU_nC_nN(double dU, int state) {
+
+  if (state == -1) {
+    _U_nC_nN_e = dU;
```

Comment:
sort of same here. Using maps could greatly improve readibilities

Reply:
Yes completely agree. Can you give me an outline of what you are thinking?